src/render.rs

```diff
@@ -69,7 +69,7 @@ where
 /// You do not need to implement this trait by hand: there is a blanket
 /// implementation for all `Render` types that fulfill the `RootRender`
 /// requirements.
-pub trait RootRender: Any + Render {
+pub trait RootRender: Any + for<'a> Render<'a> {
```

Comment:
In #46 you mentioned this is the same as `'static`, but I'm not sure it is in this case. It would be simpler without higher-order lifetimes, so I was trying to change it to `Render<'static>` or `Render + 'static`. The compiler didn't like either of those, which kinda makes sense to me.

The way I understand it, this trait bound says RootRender requires its types to also implement Any and Render for all lifetimes `'a`. `'static` is just the longest possible lifetime, so it wouldn't fulfill that requirement.

Am I missing something?

Reply:
If something is valid for all lifetimes then it must be `'static` which is the lifetime that is always valid. If it were not `'static` then it would not be valid for all lifetimes, leading to a contradiction.

The only time that `for<'a>` is useful is when you are taking some generic function and calling it with parameters that contain "unknown" lifetimes or when you wish to call it multiple times with arguments of different lifetimes.

For this case, I think `Render<'static> + 'static` should do the trick here.

Reply:
So I think we both might be right. 

Even though Render is a generic trait, it is object safe when used with higher-order lifetimes. That indicates to me that you are correct about `for<'a> Render<'a>` essentially being `'static` from the perspective of the implementation of this trait. The compiler is effectively able to assume that the lifetime of the bump arena in `RenderContext` is static since it will always be inconsequential to the functionality inside of `render`.

That being said, the bound `for<'a> Render<'a>` does seem to be required, maybe not for the callee, but the caller. Here's an example that I've boiled down:

```rust
use dodrio::{bumpalo::{self, Bump}};

struct Context<'a> {
    bump: &'a Bump
}

#[derive(Debug, PartialEq, Eq)]
struct Node<'a> {
    name: &'a str
}

trait Render<'a> {
    fn render(&self, cx: &mut Context<'a>) -> Node<'a>;
}

struct Test;

impl<'a> Render<'a> for Test {
    fn render(&self, cx: &mut Context<'a>) -> Node<'a> {
        Node { name: bumpalo::format!(in cx.bump, "Test{}", 1).into_bump_str() }
    }
}

#[cfg(test)]
mod tests {
    use super::*;


    #[test]
    fn it_works() {
        let bump = Bump::default();
        let mut cx = Context { bump: &bump  };
        let comp: Box<dyn for<'a> Render<'a>> = Box::new(Test);

        assert_eq!(Node { name: "Test1" }, comp.render(&mut cx));
    }
}
```

This test compiles and runs as expected. It's emulating the situation I currently have in my fork of dodrio.

Now, when I switch the type of comp to `Box<dyn Render<'static> + 'static>`, the compiler doesn't like this example anymore.

```rust
use dodrio::{bumpalo::{self, Bump}};

struct Context<'a> {
    bump: &'a Bump
}

#[derive(Debug, PartialEq, Eq)]
struct Node<'a> {
    name: &'a str
}

trait Render<'a> {
    fn render(&self, cx: &mut Context<'a>) -> Node<'a>;
}

struct Test;

impl<'a> Render<'a> for Test {
    fn render(&self, cx: &mut Context<'a>) -> Node<'a> {
        Node { name: bumpalo::format!(in cx.bump, "Test{}", 1).into_bump_str() }
    }
}

#[cfg(test)]
mod tests {
    use super::*;


    #[test]
    fn it_works() {
        let bump = Bump::default();
        let mut cx = Context { bump: &bump  };
        let comp: Box<dyn Render<'static> + 'static> = Box::new(Test);

        assert_eq!(Node { name: "Test1" }, comp.render(&mut cx));
    }
}
```

Compiler output:

```
   Compiling dodrio-test v0.1.0 (dodrio-test)
error[E0597]: `bump` does not live long enough
  --> src/lib.rs:32:38
   |
32 |         let mut cx = Context { bump: &bump  };
   |                                      ^^^^^ borrowed value does not live long enough
33 |         let comp: Box<dyn Render<'static> + 'static> = Box::new(Test);
   |                   ---------------------------------- type annotation requires that `bump` is borrowed for `'static`
...
36 |     }
   |     - `bump` dropped here while still borrowed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0597`.
error: could not compile `dodrio-test`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

This is because while the `Test` component may treat its `Context` as `'static`, the `it_works` function knows that `cx` has a lifetime that is shorter than `'static`. Therefore, the trait bound of `Render<'static>` cannot be satisfied.

If I'm missing something, could you show me what you mean with this example? Maybe we're talking about different things.

Reply:
Ah ok, this makes sense. Thanks for breaking it down!